Implement MathJax renderer definitions

jolyonb opened the pull request:
Ok, so I was too excited. And I needed it for the QM course I'm working on ;-)

The big piece here is the `MJxPrep.js` file that goes in the static folder. It hooks into the MathJax library to teach AsciiMath how to render a bunch of functions. Some functions are too complicated for a simple function definition and require a preprocessor (also implemented). Documentation has been updated to point out when these definitions are needed, as well as how to use them. The example course has been updated to reflect it. Pay particular attention to [the factorial function](https://edge.edx.org/courses/course-v1:MITx+grading-library+examples/courseware/examples/formulagrader/), which now behaves really beautifully! Other functions also behave as functions, and don't get incorrectly separated when you divide by them.

I've implemented fixes for all of the functions in our function list that weren't working correctly.

Version number is updated to 1.1.1.

Feedback welcome!

ChristopherChudzicki commented on docs/renderer.md:
```diff
@@ -0,0 +1,32 @@
+# MathJax Renderer Definitions
+
+When math input is expected from students, edX offers a math preview that attempts to show their expression in normal mathematical notation. This preview is rendered by MathJax via AsciiMath. While the preview does a reasonably good job, there are many situations where it falls down, even for standard edX functions (for example, try typing in `1/arctanh(x)` in an edX input box!). We have construc ...
```
Question: Does the preprocessor above work with `<textline math='true' />` AND `<formulaequationinput />`? How much of this discussion is specific to `<textline math='true' />`? For example, `<formulaequationinput />` displays `1/arctanh(x)` correctly. But it would be great if `<formulaequationinput />` could get the factorial improvements, too.

I know that MIT Physics tends to favor the `<textline />` approach, but I _believe_ Jen prefers `<formulaequationinput />`. And it seems likely that lots of people don't even know `<textline math='true' />` exists. So maybe nice to **briefly** mention the difference?

jolyonb replied:
Ah. `formulaequationinput` doesn't let us use vecx and hatx, so I've never used it. Now, `textline` uses AsciiMath, while `formulaequationinput` uses server-side parsing to generate LaTeX. So, all of these renderer things only apply to textline. I'll update the documentation appropriately.

jolyonb commented on course/problem/formula2.xml:
```diff
@@ -2,7 +2,7 @@
   <p>We have significantly expanded the sampling capabilities of our grader compared to edX. For example, we can sample integers, allowing us to use the factorial function (which is also defined in edX!). Here is an example.</p>
 
-  <p> What is the Taylor series expansion of \(\exp(x)\) about \(x = 0\)? (Answer: "x^n/(fact(n))"). Unfortunately, the math preview doesn't recognize fact() as a function properly, necessitating wrapping it in an extra set of parentheses to make it display as expected.</p>
+  <p>What is the Taylor series expansion of \(\exp(x)\) about \(x = 0\)? (Answer: "x^n/fact(n)"). Try typing "fact(n-1)" to see what happens in that case, too.</p>
```
Gamma wouldn't help here. The gamma function has poles at 0, -1, -2, etc. I will increase that range though.

ChristopherChudzicki replied:
Sure, but you could sample at real values instead, still have all the right algebraic properties, and almost never hit a pole? But not really the point of this PR :)

jolyonb replied:
Hmm. That's actually a good idea. Want to make a separate PR for that? Should be pretty easy to do.

ChristopherChudzicki replied:
Sure.

ChristopherChudzicki commented on MJxPrep.js:
```diff
@@ -0,0 +1,81 @@
+            AM.newsymbol({input:"arccoth", tag:"mi", output:"arccoth", tex:null, ttype:AM.TOKEN.UNARY, func:true});
+            AM.newsymbol({input:"re", tag:"mi", output:"Re", tex:null, ttype:AM.TOKEN.UNARY, func:true});
+            AM.newsymbol({input:"im", tag:"mi", output:"Im", tex:null, ttype:AM.TOKEN.UNARY, func:true});
+            AM.newsymbol({input:"conj", tag:"mi", output:"conj", tex:null, ttype:AM.TOKEN.UNARY, func:true});
```
I think overline would be fine with complex expressions, see below, which might make it simpler than stars. But again, no need to decide now.

<img width="798" alt="screen shot 2018-07-04 at 2 09 50 pm" src="https://user-images.githubusercontent.com/9010790/42290561-1130b826-7f94-11e8-92ee-3c0c1df06baf.png">

jolyonb replied:
Yeah ok, this works really well :-)

jolyonb commented:
Ok, I've updated all the documentation to point out that this only works for textline, fixed conj, updated the example course, and I even added a CSS selector to the refresh from the library. Let me know if you find anything else!
